persist-tool/add.c

```diff
@@ -0,0 +1,264 @@
+{
+  gchar *p = strchr(value_str, ':');
+  if (!p)
+    return -1;
+
+  p = strchr(p, '"');
+  if (!p || strlen(p)<1)
+    return -1;
+
+  p = g_strchug(p+1); // p points to the first value
+
+  gint count = 0;
+  gint token_index = 0;
+  gchar *token = NULL;
+
+  gchar **token_list = g_strsplit (p," ",-1);
+  while ((token = token_list[token_index++]))
+    {
+      int value;
+      int result = sscanf(token, "%X", &value);
+
+      if (result == EOF || result == 0)
+        continue;
+
+      if (count >= buffer_length)
+        break;
+
+      if (value > 0xFF)
+        {
+          fprintf(stderr,"invalid numeric value (%s)\n", token);
+          return -1;
+        }
+
+      buffer[count++] = (gchar)value;
+    }
+
+  g_strfreev(token_list);
+  return count;
+}
+
+void
+persist_state_entry_free(PersistStateEntry *self)
+{
+  g_free(self->name);
+  g_free(self->value);
+  g_free(self);
+}
+
+gint
+add_entry_to_persist_file(gchar *entry, PersistTool *self)
+{
+  PersistStateEntry *persist_entry = persist_state_entry_new();
+  GError *error = NULL;
+  gint result = 0;
+
+  if (!persist_state_parse_entry(persist_entry, entry))
+    {
+      result = 1;
+      fprintf(stderr, "%.*s", NAME_LENGTH, entry);
+      error = g_error_new(1, 1, "Invalid entry syntax");
+      goto exit;
+    }
+
+  gchar buffer[MAX_ENTRY_VALUE_LENGTH];
+  gint value_count = parse_value_string(persist_entry->value, buffer, MAX_ENTRY_VALUE_LENGTH);
+  if (value_count < 0)
+    {
+      result = 1;
+      fprintf(stderr,"value string is invalid (%s)\n", entry);
+      error = g_error_new(1, 1, "Invalid value string in input");
+      goto exit;
+    }
+
+  PersistEntryHandle handle = persist_state_alloc_entry(self->state, persist_entry->name, value_count);
+  if (!handle)
+    {
+      result = 1;
+      fprintf(stderr, "%.*s", NAME_LENGTH, entry);
+      error = g_error_new(1, 1, "Can't alloc entry");
+      goto exit;
+    }
+
+  gpointer block = persist_state_map_entry(self->state, handle);
+  memcpy(block, buffer, value_count);
+
+  persist_state_unmap_entry(self->state, handle);
+
+exit:
+  if (result == 0)
+    {
+      fprintf(stderr, "%s\t-->>OK\n", persist_entry->name);
+    }
+  else
+    {
+      fprintf(stderr, "%s\t-->>FAILED (error: %s)\n", persist_entry->name, error ? error->message : "<UNKNOWN>");
+    }
+  g_clear_error(&error);
+
+  persist_state_entry_free(persist_entry);
+  return result;
+}
+
+static void
+lookup_entry(gpointer data, gpointer user_data)
+{
+
+  PersistTool *self = (PersistTool *)user_data;
+  gchar *name = (gchar *)data;
+
+  gsize value_len = 0;
+  guint8 version = 0;
+  persist_state_lookup_entry(self->state, name, &value_len, &version);
+}
+
+/* this function reads all entries in persist to update the in_use struct member */
+static void
+dump_all_entries(PersistTool *self)
+{
+  GList *keys = g_hash_table_get_keys(self->state->keys);
+  g_list_foreach(keys, lookup_entry, self);
```

Comment:
As I see lookup_entry just fetches the entry handle, and ignores the result. Isn't a `fprintf` missing here?

Reply:
This is intentional. As I wrote in eralier comment, I need to call dump on all entries to set the in_use bit to 1.